### ISSUE TITLE
Functionality missing when Docker registry has different URLs

### ISSUE THREAD

**Issue by arthurdm** [appsody/appsody#410]
tl;dr: I think this could be fixed if there was a `--push` option inside `appsody build`.

When using a remote Docker registry from OKD or OpenShift we cannot do push + deployment using the Appsody CLI because the URL needed for `docker push` is different than the URL that goes into `app-deploy.yaml`'s `applicationImage` key.  

For example, in my IBM Cloud + OpenShift cluster I have to run this command to push the image:
`docker push docker-registry-default.polaris-ocp-4c666281ea4dfbebe4821a4891d29f22-0001.us-south.containers.appdomain.cloud/arthur/app:v1`

But then this URL for OpenShift to pull the image into the pod:  `docker-registry.default.svc:5000/arthur/app:v1 `

What I tried:
*  `docker login ...to remote OpenShift registry...`

*  `appsody build -t docker-registry-default.polaris-ocp-4c666281ea4dfbebe4821a4891d29f22-0001.us-south.containers.appdomain.cloud/arthur/app:v1`

    *  This builds the image with the correct tag, but it's still just sitting in my laptop's registry at this point.

At this point, I tried two things:
*  Try 1: `appsody deploy --push -t docker-registry-default.polaris-ocp-4c666281ea4dfbebe4821a4891d29f22-0001.us-south.containers.appdomain.cloud/arthur/app:v1 -n arthur`
    *  With this we get a successful `docker push`, but an `ImagePullErr` during the pod's deployment.

*  Try 2:   `appsody deploy --push -t docker-registry.default.svc:5000/arthur/app:v1 -n arthur`
    *  This fails the `docker push` because the registry URL is wrong.

So what I did to work around this bug is I manually did a docker push:
`docker push docker-registry-default.polaris-ocp-4c666281ea4dfbebe4821a4891d29f22-0001.us-south.containers.appdomain.cloud/arthur/app:v1`

Then proceeded to deploy the app (without the `--push` flag)

`appsody deploy -t docker-registry.default.svc:5000/arthur/app:v1 -n arthur`

I believe this could be solved by allow to `--push` from within `appsody build`.

**Comment by seabaylea** [comment]
FYI, Carlos Santana has hit this as well.

The cleanest solution might be to allow the .appsody-config.yaml to specify the remote repo you want to use, and the image name you want to use (getting away from repeatedly having to enter the same flags on the CLI).

**Comment by arthurdm** [comment]
that could work too @seabaylea.     From Jane's perspective, it would be nice to not have to worry about having the URL + credentials to do a `docker login` and having to know which URL to use during the push vs deployment.

Perhaps this information is something that could be pulled from a devOps GitHub repo that Todd maintained.  So then Jane just pulls down a copy of `.appsody-config.yaml` (or some other name) corresponding to the target dev cluster she wants to work with - and that file has the information that the appsody CLI needs to do a docker login, push, deploying, etc.

**Comment by chilanti** [comment]
We'll need to do some serious design sessions around this - there are a number of ramifications (and backward compatibility issues) if we go down the path that is being suggested. `appsody deploy` was primarily thought as a way to facilitate local deployment (i.e. against a cluster that is locally installed on Jane's machine). The default behavior of `appsody deploy` takes that into account and tags the image accordingly. What's being proposed changes that, in cases where certain properties (repo URLs, image name, etc.) were present in the config file.

**Comment by csantanapr** [comment]
I hit this problem with a remote cluster like an OpenShift on IBM Cloud, but also using a local cluster like crc (aka minishit4) where the docker-registry is accessible via from the docker cli differently than the image us to pull the image.

This is what I had to do with a local crc cluster ocp4.2
```
appsody deploy --push -t default-route-openshift-image-registry.apps-crc.testing/default/demo-10-03
# edit app-deploy.yaml
cat app-deploy.yaml | grep Image
  applicationImage: image-registry.openshift-image-registry.svc:5000/default/demo-10-03
# then use oc apply
oc apply -f app-deploy.yaml
appsodyapplication.appsody.dev/demo-10-03 configured
```

**Comment by MohammedFadin** [comment]
@csantanapr  Thank you. It's a workaround that wouldn't fit Jane's responsibility, to access and edit   (someone would argue with me on that).

I got a message on slack, that this issue is on top priority for fix.

**Comment by chilanti** [comment]
The Appsody team met recently to discuss possible approaches to resolve this issue.
Here is how we plan to address it:
1) Short term - we plan to add flags to `appsody build` and `appsody deploy` that allow the user to specify the "push registry URL" and the "pull registry URL". 
* `appsody build` will only support the "pull registry URL".
* `appsody deploy` will support specifying both a push and pull registry URL.

So - for example - Jane will be able to: 
`appsody deploy -t <my-repo>/<my-image> --push-url <externally addressable URL> --pull-url <internally addressable URL>`

2) Longer term - we plan to support a configuration-based mechanism that allows the developer to configure "target environments". A target environment configuration includes a specification of the push URL, and of the pull URL. Once that configuration is in place, the user (Jane) will be able to just run: 
```
appsody deploy --target <my-target> -t <my-repo>/<my-image>
```
and appsody will resolve <my-target> by retrieving the URLs of the image registry (and putting them in the appropriate files/commands).

@seabaylea @kylegc @kewegner - please add any comments. 

We'll create separate issues to address the above.

**Comment by arthurdm** [comment]
thanks @chilanti - for 1, did you mean to say that `appsody build` will only support the `push URL`?
